### ISSUE TITLE
What's the difference between ‘one2one’ mode and 'one2many' mode in data preprocessing?

### ISSUE THREAD

**Issue by zhhengcs** [memray/seq2seq-keyphrase-pytorch#7]
Hi, memray. As mentioned above, and why you do this? Thank you!

**Comment by xingdi-eric-yuan** [comment]
Hello. During teacher forcing, given an input document (e.g., paper abstract in kp20k), the model output is one key phrase; during test or policy gradient training (at each time step, the decoder takes the previously generated token as input, instead of the ground truth token), the output of model is a bunch of key phrases, generated by beam search. So basically, one2one is for negative log likelihood type of training, and one2many is used in evaluation and policy gradient types of training.

**Comment by rajeshkumargp** [comment]
Are the one2one and one2many  specified,  are similar to the diagram in the title "Forward and backward propagation on each time-step" in the link  
https://leonardoaraujosantos.gitbooks.io/artificial-inteligence/content/recurrent_neural_networks.html. 

Is it so .. ?

**Comment by xingdi-eric-yuan** [comment]
@rajeshkumargp If I understood correctly, it's not. I think it's just notions of data format.

**Comment by memray** [comment]
@zhhengcs as explained by Eric.
@rajeshkumargp one2many just means one source text (long text) corresponds to multiple target texts (keyphrases), not the structure of RNN. Our model is a seq2seq RNN, so it is same to the structure **Many to many(RNN) (Translation)** as mentioned in the referred article.

**Comment by zhhengcs** [comment]
＠memray　Thank you for your help, I still have a question that how to extract multi-word phrase using baseline methods such as tf-idf and textrank ? It seems it just rank the weight of one word,how to combine them to phrase?

**Comment by zhhengcs** [comment]
@memray 　Thank you for your help, I still have a question that how to extract multi-word phrase using baseline methods such as tf-idf and textrank ? It seems it just rank the weight of one word,how to combine them to phrase?

**Comment by memray** [comment]
If I remember it correctly, all baseline methods compute the score of phrase by averaging the scores of its words. Please check [this paper](http://www.hlt.utdallas.edu/~saidul/code.html) for more details.

**Comment by zhhengcs** [comment]
@memray Thank you very much,this tool  is good to use! I saw you implement RL training strategy in your code,do you know some paper in this direction?  I have read some paper about  summarization with RL of socher et al .ACL'17. But one sentence summarization and multi-keyphrase may have difference. Do you follow some paper or just plan to do some pioneering work?

**Comment by xingdi-eric-yuan** [comment]
@zhhengcs  If you want to go with REINFORCE, then it all depends on how you choose your reward functions. For example in [this work](https://arxiv.org/pdf/1705.02012.pdf), we tried to generate questions given documents and answers using QA datasets. For question generation, we want the questions to be 1) syntactically fluent, 2) can be answered by QA models. We used a pretrained language model as metric of fluency, and use a pretrained QA model as the metric for the 2nd criterion. So here in keyphrase generation, we can try a lot of things like, to boost up the diversity of generated keyphrases, so I tried to have a reward to penalize similar/same KPs (didn't help much, might because my implementation was bad). On the other hand, it's natural to use the metric you use to evaluate keyphrase quality (like f score here) as reward, to massage your network to get slightly better f score, although personally I think this is less interesting.
